src/model_bias_analysis.py

```diff
@@ -55,3 +58,198 @@ def load_scored_madlibs(models, scored_path=SCORED_MADLIBS_PATH,
+    confusion = confusion_matrix_counts(df, score_col, label_col, threshold)
+    actual_positives = confusion['tp'] + confusion['fn']
+    actual_negatives = confusion['tn'] + confusion['fp']
+    # True positive rate, sensitivity, recall.
+    tpr = confusion['tp'] / actual_positives
+    # True negative rate, specificity.
+    tnr = confusion['tn'] / actual_negatives
+    # False positive rate, fall-out.
+    fpr = 1 - tnr
+    # False negative rate, miss rate.
+    fnr = 1 - tpr
+    # Precision, positive predictive value.
+    precision = confusion['tp'] / (confusion['tp'] +  confusion['fp'])
+    return {
+        'tpr': tpr,
+        'tnr': tnr,
+        'fpr': fpr,
+        'fnr': fnr,
+        'precision': precision,
+        'recall': tpr,
+    }
+
+def compute_equal_error_rate(df, score_col, label_col, num_thresholds=101):
+    """Returns threshold where the false negative and false positive counts are equal."""
+    # Note: I'm not sure if this should be based on the false positive/negative
+    # *counts*, or the *rates*. However, they should be equivalent for balanced
+    # datasets.
+    thresholds = np.linspace(0, 1, num_thresholds)
+    min_threshold = None
+    min_confusion_matrix = None
+    min_diff = float('inf')
+    for threshold in thresholds:
+        confusion_matrix = confusion_matrix_counts(df, score_col, label_col,
+                                                   threshold)
+        difference = abs(confusion_matrix['fn'] - confusion_matrix['fp'])
+        if difference <= min_diff:
+            min_diff = difference
+            min_confusion_matrix = confusion_matrix
+            min_threshold = threshold
+        else:
+            # min_diff should be monotonically non-decreasing, so once it
+            # increases we can break. Yes, we could do a binary search instead.
+            break
+    return {
+        'threshold': min_threshold,
+        'confusion_matrix': min_confusion_matrix,
+    }
+
+def per_model_eer(dataset, label_col, model_names, num_eer_thresholds=101):
+    """Computes the equal error rate for every model on the given dataset."""
+    model_name_to_eer = {}
+    for model_name in model_names:
+        eer = compute_equal_error_rate(dataset, model_name, label_col,
+                                       num_eer_thresholds)
+        model_name_to_eer[model_name] = eer['threshold']
+    return model_name_to_eer
+
+def per_term_negative_rates(df, terms, model_families, threshold, text_col,
+                            label_col):
+    """Computes per-term true/false negative rates for all model families.
+
+    Args:
+      df: dataset to compute rates on.
+      terms: negative rates are computed on subsets of the dataset containing
+          each term.
+      text_col: column in df containing the text.
+      label_col: column in df containing the boolean label.
+      model_families_names: list of model families; each model family is a list
```

Comment:
I think this should just be "model_families"

Reply:
Good catch. Yeah, I forgot to delete the model_family_name from the notebook.